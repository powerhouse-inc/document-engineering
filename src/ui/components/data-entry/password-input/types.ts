import type { InputBaseProps, WithDifference } from '../../../../scalars/components/types.js'

type PasswordInputWithDifference = Omit<WithDifference<string>, 'diffMode'>

interface PasswordInputProps
  extends Omit<
      React.InputHTMLAttributes<HTMLInputElement>,
      keyof InputBaseProps<string> | 'autoCorrect' | 'autoCapitalize' | 'spellCheck'
    >,
    InputBaseProps<string>,
    PasswordInputWithDifference {
  autoCorrect?: never
  autoCapitalize?: never
  spellCheck?: never
  // TODO: Implement commented props
  // requireUppercase?: boolean
  // requireLowercase?: boolean
  // requireNumbers?: boolean
  // requireSpecialCharacters?: boolean
  // disallowCommonPasswords?: boolean
  showPasswordStrength?: boolean
}

export type { PasswordInputProps }
