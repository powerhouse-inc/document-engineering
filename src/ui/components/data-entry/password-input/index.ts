export { PasswordInput } from './password-input.js'
export type { PasswordInputProps } from './types.js'
