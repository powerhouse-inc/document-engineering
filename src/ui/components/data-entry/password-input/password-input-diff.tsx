import { FormGroup, FormLabel } from '../../../../scalars/components/fragments/index.js'
import { SplittedInputDiff } from '../input/splitted-input-diff.js'
import type { WithDifference } from '../../../../scalars/components/types.js'
import type { PasswordInputProps } from './types.js'

interface PasswordInputDiffProps extends Omit<WithDifference<string>, 'diffMode'> {
  value: string
  label: PasswordInputProps['label']
  required: PasswordInputProps['required']
}

const PasswordInputDiff = ({ value = '', label, required, viewMode, baseValue = '' }: PasswordInputDiffProps) => {
  return (
    <FormGroup>
      {label && (
        <FormLabel disabled={true} required={required}>
          {label}
        </FormLabel>
      )}
      <SplittedInputDiff
        baseValue={baseValue}
        value={value}
        viewMode={viewMode}
        diffMode="sentences"
        data-testid="password-input-diff"
      />
    </FormGroup>
  )
}

export { PasswordInputDiff }
