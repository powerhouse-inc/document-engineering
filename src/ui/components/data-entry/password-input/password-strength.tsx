import { useEffect, useMemo, useState } from 'react'
import { debounce, zxcvbnAsync, zxcvbnOptions, type ZxcvbnResult } from '@zxcvbn-ts/core'
import * as zxcvbnCommonPackage from '@zxcvbn-ts/language-common'
import * as zxcvbnEnPackage from '@zxcvbn-ts/language-en'
import { cn } from '../../../../scalars/lib/index.js'
import { ProgressBar } from '../../../../scalars/components/fragments/progress-bar/index.js'
import { FormMessage } from '../../../../scalars/components/fragments/form-message/index.js'
import {
  adjustStrengthScore,
  getPasswordRequirements,
  strengthColors,
  strengthLabels,
  strengthValues,
} from './utils.js'

const options = {
  dictionary: {
    ...zxcvbnCommonPackage.dictionary,
    ...zxcvbnEnPackage.dictionary,
  },
  graphs: zxcvbnCommonPackage.adjacencyGraphs,
  useLevenshteinDistance: true,
}
zxcvbnOptions.setOptions(options)

const PasswordStrength = ({ password }: { password: string }) => {
  const [result, setResult] = useState<ZxcvbnResult | undefined>(undefined)
  const [wasEmpty, setWasEmpty] = useState(password === '')

  const debouncedZxcvbn = useMemo(
    () =>
      debounce(async (passwordToCheck: string) => {
        try {
          const response = await zxcvbnAsync(passwordToCheck)
          setResult(response)
        } catch (error) {
          console.error('Error calculating password strength: ', error)
          setResult(undefined)
        }
      }, 200),
    []
  )

  useEffect(() => {
    if (password === '') {
      setWasEmpty(true)
    }
    debouncedZxcvbn(password)
  }, [password, debouncedZxcvbn])

  useEffect(() => {
    if (password !== '' && result !== undefined) {
      setWasEmpty(false)
    }
  }, [password, result])

  const score = adjustStrengthScore(password, result?.score)
  const showWeak = password === '' || (score === undefined && wasEmpty)
  const requirements = getPasswordRequirements(password)

  return (
    <div className={cn('w-full flex flex-col gap-4 px-4 pt-2 pb-4 rounded-md')}>
      <div className={cn('w-full flex flex-col gap-2')}>
        <h3 className={cn('text-sm font-semibold text-gray-900')}>Password strength</h3>

        <ProgressBar
          value={score !== undefined ? strengthValues[score] : showWeak ? strengthValues[0] : 0}
          className={cn(
            score !== undefined
              ? strengthColors[score]
              : showWeak
                ? strengthColors[0]
                : '[&_[data-slot=progress-bar-indicator]]:bg-transparent'
          )}
        />

        <span className={cn('h-4 text-xs font-medium text-gray-600')}>
          {score !== undefined ? strengthLabels[score] : showWeak ? strengthLabels[0] : ''}
        </span>
      </div>

      <hr className={cn('w-full border-t border-gray-300')} />

      <div className={cn('w-full flex flex-col gap-4')}>
        <h4 className={cn('text-sm leading-[18px] font-medium text-gray-500')}>Suggestions</h4>
        <ul className={cn('flex flex-col gap-2')}>
          {requirements.map((requirement) => (
            <FormMessage
              key={requirement.text}
              as="li"
              className={cn(
                'relative pl-4',
                'before:absolute before:left-0 before:top-[0.4em] before:size-2 before:rounded-full',
                'transition-colors duration-200 ease-in-out',
                'before:transition-colors before:duration-200 before:ease-in-out',
                requirement.isValid ? 'text-green-900 before:bg-green-900' : 'text-gray-500 before:bg-gray-500'
              )}
            >
              {requirement.text}
            </FormMessage>
          ))}
        </ul>
      </div>
    </div>
  )
}

export { PasswordStrength }
