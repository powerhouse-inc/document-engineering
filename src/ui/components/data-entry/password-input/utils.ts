export const specialCharacters = '" !#$%&\'()*+,-./:;<?>@=[\\]^_`{|}~'

export const strengthColors = [
  '[&_[data-slot=progress-bar-indicator]]:bg-red-900',
  '[&_[data-slot=progress-bar-indicator]]:bg-red-900',
  '[&_[data-slot=progress-bar-indicator]]:bg-yellow-900',
  '[&_[data-slot=progress-bar-indicator]]:bg-yellow-900',
  '[&_[data-slot=progress-bar-indicator]]:bg-green-900',
]
export const strengthLabels = ['Weak', 'Weak', 'Medium', 'Medium', 'Strong']
export const strengthValues = [33, 33, 66, 66, 100]
