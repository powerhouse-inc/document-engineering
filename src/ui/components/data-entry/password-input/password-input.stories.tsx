import type { Meta, StoryObj } from '@storybook/react'
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from '../../../../scalars/lib/storybook-arg-types.js'
import { PasswordInput } from './password-input.js'

/**
 * The `PasswordInput` component provides an input field for password entry.
 *
 * Features include:
 * - Secure password masking with toggle visibility
 * - Password strength meter with suggestions display
 * - Character requirements (uppercase, lowercase, numbers, special characters)
 * - Common password detection and prevention
 *
 * > **Note:** This component does not have built-in validation. If you need built-in validation
 * > you can use the [PasswordField](?path=/docs/scalars-password-field--readme)
 * > component.
 */

const meta: Meta<typeof PasswordInput> = {
  title: 'Data Entry/Password Input',
  component: PasswordInput,
  decorators: [
    (Story) => (
      <div style={{ width: '280px', margin: '1rem auto 0' }}>
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: 'padded',
  },
  tags: ['autodocs'],
  argTypes: {
    ...getDefaultArgTypes(),
    ...PrebuiltArgTypes.placeholder,
    ...PrebuiltArgTypes.minLength,
    ...PrebuiltArgTypes.maxLength,
    ...getValidationArgTypes({
      enabledArgTypes: {
        validators: false,
        showErrorOnBlur: false,
        showErrorOnChange: false,
      },
    }),

    // TODO: Implement commented props
    /* requireUppercase: {
      control: 'boolean',
      description: 'The field value requires at least one capital letter',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'true' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    requireLowercase: {
      control: 'boolean',
      description: 'The field value requires at least one lowercase letter',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'true' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    requireNumbers: {
      control: 'boolean',
      description: 'The field value requires at least one number',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'true' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    requireSpecialCharacters: {
      control: 'boolean',
      description:
        'The field value requires at least one special character (from the list between double quotes): " !"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~"',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'true' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    disallowCommonPasswords: {
      control: 'boolean',
      description: 'Disallows commonly used or compromised passwords as valid input in the field',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'true' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    }, */

    showPasswordStrength: {
      control: 'boolean',
      description: 'Displays a password strength meter',
      table: {
        type: { summary: 'boolean' },
        defaultValue: { summary: 'true' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    ...PrebuiltArgTypes.viewMode,
    ...PrebuiltArgTypes.baseValue,
  },
  args: {
    name: 'password-input',
  },
} satisfies Meta<typeof PasswordInput>

export default meta

type Story = StoryObj<typeof PasswordInput>

export const Default: Story = {
  args: {
    label: 'Password input',
    placeholder: 'Password',
  },
}

export const Filled: Story = {
  args: {
    label: 'Password input',
    placeholder: 'Password',
    defaultValue: 'H0l4.mundo',
  },
}
