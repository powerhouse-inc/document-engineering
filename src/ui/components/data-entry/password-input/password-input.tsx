import React, { useCallback, useEffect, useRef, useState } from 'react'
import { Input } from '../input/index.js'
import { cn, useUniqueId } from '../../../../scalars/lib/index.js'
import { FormDescription } from '../../../../scalars/components/fragments/form-description/index.js'
import { FormGroup } from '../../../../scalars/components/fragments/form-group/index.js'
import { FormLabel } from '../../../../scalars/components/fragments/form-label/index.js'
import { FormMessageList } from '../../../../scalars/components/fragments/form-message/index.js'
import { Popover, PopoverAnchor, PopoverContent } from '../../../../scalars/components/fragments/popover/index.js'
import { PasswordStrength } from './password-strength.js'
import { Icon } from '../../icon/index.js'
import type { PasswordInputProps } from './types.js'

const PasswordInput = React.forwardRef<HTMLInputElement, PasswordInputProps>(
  (
    {
      id: idProp,
      className,
      label,
      description,
      value,
      defaultValue,
      onChange,
      disabled,
      required,
      errors,
      warnings,
      pattern,
      showPasswordStrength = true,
      showPasswordStrengthOpen = false,
      viewMode = 'edition',
      // TODO: Implement diffs
      // baseValue,
      ...props
    },
    ref
  ) => {
    const prefix = useUniqueId()
    const id = idProp ?? `${prefix}-password-input`

    const [isPopoverOpen, setIsPopoverOpen] = useState(showPasswordStrengthOpen)
    const [password, setPassword] = useState(value ?? defaultValue ?? '')
    const [showPassword, setShowPassword] = useState(false)

    const containerRef = useRef<HTMLDivElement | null>(null)

    const handleChange = useCallback(
      (event: React.ChangeEvent<HTMLInputElement>) => {
        setIsPopoverOpen(true)
        setPassword(event.target.value)
        onChange?.(event)
      },
      [onChange]
    )

    useEffect(() => {
      if (value !== undefined) {
        setPassword(value)
      }
    }, [value])

    const hasWarning = Array.isArray(warnings) && warnings.length > 0
    const hasError = Array.isArray(errors) && errors.length > 0

    // TODO: Implement PasswordInputDiff
    if (viewMode !== 'edition') {
      return null
    }

    return (
      <FormGroup>
        {!!label && (
          <FormLabel
            htmlFor={id}
            disabled={disabled}
            hasError={hasError}
            required={required}
            onClick={(e) => {
              e.preventDefault()
              ;(e.target as HTMLLabelElement).control?.focus()
            }}
          >
            {label}
          </FormLabel>
        )}
        <Popover open={isPopoverOpen} onOpenChange={setIsPopoverOpen}>
          <PopoverAnchor asChild={true}>
            <div ref={containerRef} className={cn('relative')}>
              <Input
                ref={ref}
                id={id}
                value={password}
                onChange={handleChange}
                onClick={(e) => {
                  setIsPopoverOpen(true)
                  props.onClick?.(e)
                }}
                className={cn('pr-9', className)}
                disabled={disabled}
                aria-invalid={hasError}
                aria-label={!label ? 'Password input' : undefined}
                aria-required={required}
                type={showPassword ? 'text' : 'password'}
                pattern={pattern?.source}
                autoCorrect="off"
                autoCapitalize="off"
                spellCheck={false}
                {...props}
              />
              <div className={cn('absolute right-3 top-1/2 flex size-4 -translate-y-1/2 items-center')}>
                <button
                  type="button"
                  aria-label={showPassword ? 'Hide password' : 'Show password'}
                  onClick={() => {
                    setShowPassword(!showPassword)
                  }}
                  className={cn('group focus-visible:outline-none')}
                >
                  <Icon
                    name={showPassword ? 'Hide' : 'Show'}
                    size={16}
                    className={cn(
                      'text-gray-500 group-hover:text-gray-700 group-focus:text-gray-700 pointer-events-none'
                    )}
                    aria-hidden="true"
                  />
                </button>
              </div>
            </div>
          </PopoverAnchor>
          <PopoverContent
            align="start"
            onOpenAutoFocus={(e) => {
              e.preventDefault()
            }}
            onInteractOutside={(e) => {
              if (e.target instanceof Element && containerRef.current) {
                const isWithinContainer = containerRef.current.contains(e.target as Node)
                if (isWithinContainer) {
                  e.preventDefault()
                }
              }
            }}
            className={cn('border-none shadow-[0px_2px_12px_0px_rgba(37,42,52,0.10)]')}
          >
            {showPasswordStrength && <PasswordStrength password={password} />}
          </PopoverContent>
        </Popover>
        {!!description && <FormDescription>{description}</FormDescription>}
        {hasWarning && <FormMessageList messages={warnings} type="warning" />}
        {hasError && <FormMessageList messages={errors} type="error" />}
      </FormGroup>
    )
  }
)

PasswordInput.displayName = 'PasswordInput'

export { PasswordInput }
