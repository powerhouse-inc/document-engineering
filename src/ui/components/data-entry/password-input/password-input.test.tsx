import { render, screen, waitFor } from '@testing-library/react'
import { describe, expect, it, vi } from 'vitest'
import { PasswordInput } from './password-input.js'

describe('PasswordInput Component', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<PasswordInput name="password" label="Password input" placeholder="Password" />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render with label', () => {
    render(<PasswordInput name="password" label="Password input" />)
    expect(screen.getByText('Password input')).toBeInTheDocument()
  })

  it('should render with description', () => {
    render(<PasswordInput name="password" label="Password input" description="Password input description" />)
    expect(screen.getByText('Password input description')).toBeInTheDocument()
  })

  it('should handle disabled state', () => {
    render(<PasswordInput name="password" label="Password input" disabled />)
    expect(screen.getByLabelText('Password input')).toBeDisabled()
  })

  it('should display error messages', async () => {
    render(<PasswordInput name="password" label="Password input" errors={['Invalid password']} />)
    await waitFor(() => {
      expect(screen.getByText('Invalid password')).toBeInTheDocument()
    })
  })

  it('should display warning messages', () => {
    render(
      <PasswordInput
        name="password"
        label="Password input"
        warnings={['Please change your password as soon as possible']}
      />
    )
    expect(screen.getByText('Please change your password as soon as possible')).toBeInTheDocument()
  })

  it('should have correct ARIA attributes', async () => {
    render(<PasswordInput name="password" label="Password input" required errors={['Invalid password']} />)

    const input = screen.getByLabelText(/Password input/)
    expect(input).toHaveAttribute('aria-required', 'true')
    await waitFor(() => {
      expect(input).toHaveAttribute('aria-invalid', 'true')
    })
  })

  it('should not invoke onChange on mount when it has a defaultValue', () => {
    const onChange = vi.fn()
    render(<PasswordInput name="password" label="Password input" defaultValue="H0l4.mundo" onChange={onChange} />)
    expect(onChange).not.toHaveBeenCalled()
  })
})
