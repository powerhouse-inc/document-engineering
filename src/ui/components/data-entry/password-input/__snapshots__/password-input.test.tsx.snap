// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PasswordInput Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
      for=":r0:-password-input"
    >
      Password input
    </label>
    <div
      class="relative"
    >
      <input
        aria-invalid="false"
        autocapitalize="off"
        autocorrect="off"
        class="flex h-9 w-full rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-2 font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 pr-9"
        id=":r0:-password-input"
        name="password"
        placeholder="Password"
        spellcheck="false"
        type="password"
        value=""
      />
      <div
        class="absolute right-3 top-1/2 flex size-4 -translate-y-1/2 items-center"
      >
        <button
          aria-label="Show password"
          class="group focus-visible:outline-none"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="text-gray-500 group-hover:text-gray-700 group-focus:text-gray-700 pointer-events-none"
            data-testid="icon-fallback"
            fill="currentcolor"
            style="width: 16px; height: 16px;"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M7.99999 11.7667C5.10282 11.7667 2.66271 10.282 1.1291 8.00001C2.66271 5.71801 5.10282 4.23333 7.99999 4.23333C10.8972 4.23333 13.3373 5.71801 14.8709 8.00001C13.3373 10.282 10.8972 11.7667 7.99999 11.7667ZM7.99999 3.23333C4.60785 3.23333 1.78972 5.04637 0.109325 7.73501C0.00799137 7.89715 0.00799179 8.10287 0.109327 8.26501C1.78972 10.9536 4.60785 12.7667 7.99999 12.7667C11.3921 12.7667 14.2103 10.9536 15.8907 8.26501C15.992 8.10287 15.992 7.89715 15.8907 7.73501C14.2103 5.04637 11.3921 3.23333 7.99999 3.23333ZM7.99999 10.1333C9.1782 10.1333 10.1333 9.1782 10.1333 8C10.1333 6.82179 9.1782 5.86666 7.99999 5.86666C6.82178 5.86666 5.86666 6.82179 5.86666 8C5.86666 9.1782 6.82178 10.1333 7.99999 10.1333Z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
